spt/features/vag_trace.cpp

```diff
@@ -0,0 +1,273 @@
+#include "stdafx.h"
+
+#if defined(SSDK2007) || defined(SSDK2013)
+#include "..\feature.hpp"
+#include "icliententity.h"
+#include "..\utils\game_detection.hpp"
+#include "ent_utils.hpp"
+#include "interfaces.hpp"
+#include "..\cvars.hpp"
+#include "tickrate.hpp"
+#include "signals.hpp"
+#include "playerio.hpp"
+
+extern IClientEntity* getPortal(const char* arg, bool verbose);
+extern IClientEntity* GetLinkedPortal(IClientEntity* portal);
+
+ConVar y_spt_vag_trace("y_spt_vag_trace", "0", FCVAR_CHEAT, "Draws VAG trace.\n");
+ConVar y_spt_vag_target("y_spt_vag_target", "0", FCVAR_CHEAT, "Draws VAG target trace.\n");
+
+// VAG finding tool
+class VagTrace : public FeatureWrapper<VagTrace>
+{
+public:
+	void DrawTrace();
+
+	Vector CalculateAG(Vector enter_origin, QAngle enter_angles, Vector exit_origin, QAngle exit_angles);
+
+	void SetTarget(Vector target);
+
+	Vector ReverseAG(Vector enter_origin, QAngle enter_angles, QAngle exit_angles);
+
+protected:
+	virtual bool ShouldLoadFeature() override;
+
+	virtual void LoadFeature() override;
+
+	virtual void UnloadFeature() override;
+
+private:
+	Vector target = {0.0f, 0.0f, 0.0f};
+};
+
+VagTrace spt_vag_trace;
+
+CON_COMMAND(y_spt_vag_target_set, "Set VAG target\n")
+{
+	spt_vag_trace.SetTarget(spt_playerio.GetPlayerEyePos());
+}
+
+void VagTrace::DrawTrace()
+{
+	bool vag_trace = y_spt_vag_trace.GetBool();
+	bool vag_target = y_spt_vag_target.GetBool();
+	if (!vag_trace && !vag_target)
+		return;
+
+	float lifeTime = spt_tickrate.GetTickrate() * 2;
+	if (vag_target)
+	{
+		interfaces::debugOverlay->AddBoxOverlay(target,
+		                                        Vector(-20, -20, -20),
+		                                        Vector(20, 20, 20),
+		                                        QAngle(0, 0, 0),
+		                                        255,
+		                                        255,
+		                                        0,
+		                                        100,
+		                                        lifeTime);
+	}
+
+	auto enter_portal = getPortal(_y_spt_overlay_portal.GetString(), false);
```

Comment:
it's slightly confusing that the overlay portal determines how the vag searcher works, but I guess most of the time you would want them to match